doc/yaml_config.md

```diff
@@ -0,0 +1,54 @@
+YaML Config
+-----------
+
+Why YaML
+========
+
+With our previous experience we discard at the start XML due to its poor readability.
+JSON format we also consider has serious disadvantage - lack of comments.
+So we decide to go with YaML which is still comfortable to read and also machine read.
+Only disadvantage we found is lack of mature schema support for YaML.
+
+Rules for Config
+================
+
+1. No default values in code. All values have to be defined in yaml config.
+2. No objects, just built-in data types.
+
+Structure
+=========
+
+Top level config element is map. Each key is described here in following sections.
+
+### software
+
+Software manager related options. It is map with following keys:
+
+#### installation\_repositories
+
+Array of url for installation repositories.
+
+### distros
+
+List of supported distros that can be offered in installer.
+
+### conditions
```

Comment:
Some tools like [codefresh](https://codefresh.io/docs/docs/codefresh-yaml/condition-expression-syntax/) use a slightly different approach for conditions, for example:

~~~yaml
partitioning:
  lvm: true
  disks: [/dev/sda]
  when:
    arch: x86
    dist: TW

partitioning:
  lvm: true
  disks: [/dev/sda]
  when:
    arch: s390
~~~

 This avoid the problem of having nested conditions in *#value* as you mentioned.

Reply:
Well, it means that we need to process that when everywhere. On other hand here it will be if condition is satisfied, it will overwrite default. And I also hope it will be more readable to have
```yaml
partitioning:
  lvm: true
  disks: [/dev/sda]

conditions:
  - arch: aarch64
  - value:
    partitioning:
       lvm: false
```

so it should be clearly visible what will change on specific architectures or distros.

Reply:
@joseivanlopez btw I think your example won't work as you use same key name for hash. So it has to be like `tw_partitioning` and `s390_partitioning`

Reply:
How that *conditions* scale when there are more than one set of conditions? Let's say we have to apply different options for: a) x86 and b) x86 and TW. Something like:

~~~yaml
partitioning:
  lvm: true
  when:
    all:
      arch: x86
    value:
      lvm: false
      when:
        all:
          dist: TW
        value:
          lvm: true
~~~

Reply:
it is order dependent. So if you want to have different ones, then use at first the generic one and later specific one. I think it is quite commonly used in configuration. Basically I see there three different strategies:

1. stop on first matched conditions.
2. use the last match
3. apply all matched configs in order

For me advantage of 3 is that it does not restrict use to repeat some settings. So e.g. when we have condition for network based on arch, then software based on distro and then partitioning based on distro and arch, we can easily split it to different sections where it makes sense and really have it separated as needed.